### ISSUE TITLE
Customize format of import/export buttons

### ISSUE THREAD

**Issue by LeaVerou** [mavoweb/mavo#477]
@karger brought this up at some point.
We have import/export buttons, but they always deal with JSON. It would be nice if there was a way to customize what formats they import and export (perhaps even have multiple instances of them, each exporting a different format, e.g. one Export JSON button, one Export CSV button and so on).

Ideally, IMO the most readable syntax would be just using an identifier after the button name, e.g. `mv-bar="export export-csv` (where `export` and `export-json` are equivalent). However, this means we need to introduce the concept of arbitrary parameters for Mavo bar buttons so it's a more substantial change. Not a bad one though!

**Comment by karger** [comment]
since import and export are generally reading/writing a file, why not 
stick with the file-chooser metaphor where filetype (and name) are 
defaulted but overridden in a dialog box?   export/import are rare 
enough that a dialog seems fine.

I don't see a reason for the author to need or want to specify a 
restricted set of file types.  I think the dialog should always offer 
all the types mavo can manage.  Most convenient for the user.

There may be good use cases for copy/paste data as well, ie 
import/export via the clipboard, and/or through a popup text box.

------ Original Message ------

**Comment by LeaVerou** [comment]
I ...don't think this is possible in the web platform :/ 
Unless you mean a custom dialog, not the OS-provided one. That we can do. Whether it's a better way is debatable. Perhaps we could do that for parameter-less import/export and *also* have import-xxx/export-xxx for more targeted buttons?

**Comment by karger** [comment]
yes, i was suggesting using a custom dialog, but one in the style of the 
os-provided one.

------ Original Message ------

**Comment by joyously** [comment]
The formats are provided by plugins, aren't they?

**Comment by LeaVerou** [comment]
@joyously Not all of them, a few are built-in. But Mavo can still iterate over "installed" formats, even if they are provided by plugins.

**Comment by joyously** [comment]
OK, I just didn't know when they would be loaded. I didn't think format plugins would be loaded unless needed (hoping so, for performance), but adding it as a user choice means you don't know what is needed until right when you need it. So I guess the list would be populated by some static list of formats?

**Comment by LeaVerou** [comment]
Right now format plugins are loaded from the get-go, since they could be needed to read the stored data in.
